File: android/src/main/java/com/risemaxi/graft/GraftPatch.java
package com.risemaxi.graft;

import androidx.annotation.NonNull;
import com.github.luben.zstd.ZstdInputStream;
import com.risemaxi.graft.classes.Manifest;
import com.risemaxi.graft.classes.ManifestFile;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.nio.ByteBuffer;
import java.security.MessageDigest;
import java.util.HashMap;
import java.util.Map;
import org.json.JSONArray;
import org.json.JSONObject;

/**
 * Applies a patch archive produced by `tools/make-patch.mjs`.
 *
 * <p>The plan inside the archive says how to reconstruct each file; it never says whether the result
 * is acceptable. Reconstruction is driven by the signed manifest's file list, every output file is
 * verified against its manifest digest, and a manifest entry the plan does not cover is an error.
 */
final class GraftPatch {

    interface BaseReader {
        @NonNull
        byte[] read(@NonNull String href) throws Exception;
    }

    private static final byte[] MAGIC = { 'G', 'R', 'A', 'F', 'T', 'P', '1', '\n' };
    private static final int SCHEMA = 1;
    private static final int WINDOW_LOG_MAX = 27;

    private GraftPatch() {}

    static void apply(@NonNull File archive, @NonNull BaseReader base, @NonNull Manifest manifest, @NonNull File targetDirectory)
        throws Exception {
        byte[] container = inflate(archive);
        ByteBuffer buffer = ByteBuffer.wrap(container);

        for (byte expected : MAGIC) {
            if (!buffer.hasRemaining() || buffer.get() != expected) {
                throw new Exception(GraftPlugin.ERROR_PATCH_FAILED);
            }
        }

        JSONObject plan = new JSONObject(new String(readBlock(buffer), "UTF-8"));
        if (plan.optInt("schema") != SCHEMA) {
            throw new Exception(GraftPlugin.ERROR_PATCH_FAILED);
        }

        int payloadCount = readLength(buffer);
        byte[][] payloads = new byte[payloadCount][];
        for (int index = 0; index < payloadCount; index += 1) {
            payloads[index] = readBlock(buffer);
        }
        if (buffer.hasRemaining()) {
            throw new Exception(GraftPlugin.ERROR_PATCH_FAILED);
        }

        Map<String, JSONObject> operationByHref = indexOperations(plan.getJSONArray("ops"));

        for (ManifestFile file : manifest.getFiles()) {
            JSONObject operation = operationByHref.get(file.getHref());
            if (operation == null) {
                throw new Exception(GraftPlugin.ERROR_PATCH_FAILED);
            }

            byte[] content;
            String kind = operation.getString("op");
            switch (kind) {
                case "keep":
                    content = base.read(safeHref(operation.getString("from")));
                    break;
                case "patch":
                    content = patch(
                        base.read(safeHref(operation.getString("from"))),
                        payloadAt(payloads, operation.getInt("payload")),
                        file.getSize()
                    );
                    break;
                case "add":
                    content = payloadAt(payloads, operation.getInt("payload"));
                    break;
                default:
                    throw new Exception(GraftPlugin.ERROR_PATCH_FAILED);
            }

            if (!digestOf(content).equals(file.getSha256())) {
                throw new Exception(GraftPlugin.ERROR_PATCH_CHECKSUM_MISMATCH);
            }

            File destination = new File(targetDirectory, file.getHref());
            File parent = destination.getParentFile();
            if (parent != null && !parent.exists() && !parent.mkdirs()) {
                throw new Exception(GraftPlugin.ERROR_PATCH_FAILED);
            }
            try (FileOutputStream out = new FileOutputStream(destination)) {
                out.write(content);
            }
        }
    }

    @NonNull
    private static Map<String, JSONObject> indexOperations(@NonNull JSONArray operations) throws Exception {
        Map<String, JSONObject> byHref = new HashMap<>();
        for (int index = 0; index < operations.length(); index += 1) {
            JSONObject operation = operations.getJSONObject(index);
            if ("delete".equals(operation.optString("op"))) {
                continue;
            }
            byHref.put(operation.getString("href"), operation);
        }
        return byHref;
    }

    /**
     * `--patch-from` emits a frame that references the base as a raw content dictionary, so applying
     * one is ordinary decompression with a prefix loaded — no patch library is involved.
     */
    @NonNull
    private static byte[] patch(@NonNull byte[] base, @NonNull byte[] delta, long size) throws Exception {
        if (size < 0 || size > Integer.MAX_VALUE) {
            throw new Exception(GraftPlugin.ERROR_PATCH_FAILED);
        }
        try (ZstdInputStream stream = new ZstdInputStream(new ByteArrayInputStream(delta))) {
            stream.setLongMax(WINDOW_LOG_MAX);
            stream.setDict(base);
            byte[] content = readFully(stream, (int) size);
            if (content.length != size) {
                throw new Exception(GraftPlugin.ERROR_PATCH_FAILED);
            }
            return content;
        }
    }

    @NonNull
    private static byte[] inflate(@NonNull File archive) throws Exception {
        try (FileInputStream file = new FileInputStream(archive); ZstdInputStream stream = new ZstdInputStream(file)) {
            stream.setLongMax(WINDOW_LOG_MAX);
            return readFully(stream);
        }
    }

    @NonNull
    private static byte[] readFully(@NonNull InputStream stream) throws Exception {
        return readFully(stream, 8192);
    }

    @NonNull
    private static byte[] readFully(@NonNull InputStream stream, int expectedSize) throws Exception {
        ByteArrayOutputStream out = new ByteArrayOutputStream(Math.max(expectedSize, 8192));
        byte[] chunk = new byte[8192];
        int read;
        while ((read = stream.read(chunk)) != -1) {
            out.write(chunk, 0, read);
        }
        return out.toByteArray();
    }

    @NonNull
    private static byte[] payloadAt(@NonNull byte[][] payloads, int index) throws Exception {
        if (index < 0 || index >= payloads.length) {
            throw new Exception(GraftPlugin.ERROR_PATCH_FAILED);
        }
        return payloads[index];
    }

    private static int readLength(@NonNull ByteBuffer buffer) throws Exception {
        if (buffer.remaining() < 4) {
            throw new Exception(GraftPlugin.ERROR_PATCH_FAILED);
        }
        int length = buffer.getInt();
        if (length < 0 || length > buffer.remaining()) {
            throw new Exception(GraftPlugin.ERROR_PATCH_FAILED);
        }
        return length;
    }

    @NonNull
    private static byte[] readBlock(@NonNull ByteBuffer buffer) throws Exception {
        byte[] block = new byte[readLength(buffer)];
        buffer.get(block);
        return block;
    }

    /**
     * The plan is untrusted input, so a base href it names is held to the same rule as a manifest
     * href: relative, and no segment that could climb out of the bundle directory.
     */
    @NonNull
    private static String safeHref(@NonNull String href) throws Exception {
        if (href.isEmpty() || href.startsWith("/")) {
            throw new Exception(GraftPlugin.ERROR_PATCH_FAILED);
        }
        for (String segment : href.split("/")) {
            if (segment.isEmpty() || segment.equals(".") || segment.equals("..")) {
                throw new Exception(GraftPlugin.ERROR_PATCH_FAILED);
            }
        }
        return href;
    }

    @NonNull
    private static String digestOf(@NonNull byte[] content) throws Exception {
        byte[] digest = MessageDigest.getInstance("SHA-256").digest(content);
        StringBuilder hex = new StringBuilder(digest.length * 2);
        for (byte value : digest) {
            hex.append(Character.forDigit((value >> 4) & 0xf, 16));
            hex.append(Character.forDigit(value & 0xf, 16));
        }
        return hex.toString();
    }
}

File: ios/Sources/RisemaxiGraft/GraftPatch.swift
import CryptoKit
import Foundation
import libzstd

/// The plan says how to reconstruct each file, never whether the result is acceptable: drive this from
/// the signed manifest's file list, not from the plan.
enum GraftPatch {
    private static let magic: [UInt8] = Array("GRAFTP1\n".utf8)
    private static let schema = 1
    private static let windowLogMax: Int32 = 27

    static func apply(archive: URL, base: URL, manifest: Manifest, to directory: URL) throws {
        let container = try decompress([UInt8](try Data(contentsOf: archive)), prefix: nil, expectedSize: nil)
        var offset = 0

        guard container.count >= magic.count, Array(container[0..<magic.count]) == magic else {
            throw CustomError.patchFailed
        }
        offset = magic.count

        let planBytes = try readBlock(container, &offset)
        guard
            let plan = try JSONSerialization.jsonObject(with: Data(planBytes)) as? [String: Any],
            plan["schema"] as? Int == schema,
            let operations = plan["ops"] as? [[String: Any]]
        else {
            throw CustomError.patchFailed
        }

        let payloadCount = try readLength(container, &offset)
        var payloads = [[UInt8]]()
        payloads.reserveCapacity(payloadCount)
        for _ in 0..<payloadCount {
            payloads.append(try readBlock(container, &offset))
        }
        guard offset == container.count else {
            throw CustomError.patchFailed
        }

        var operationByHref = [String: [String: Any]]()
        for operation in operations {
            guard let href = operation["href"] as? String, operation["op"] as? String != "delete" else {
                continue
            }
            operationByHref[href] = operation
        }

        for file in manifest.files {
            guard let operation = operationByHref[file.href], let kind = operation["op"] as? String else {
                throw CustomError.patchFailed
            }

            let content: [UInt8]
            switch kind {
            case "keep":
                content = try readBase(base, operation)
            case "patch":
                content = try decompress(
                    try payload(payloads, operation),
                    prefix: try readBase(base, operation),
                    expectedSize: file.size
                )
            case "add":
                content = try payload(payloads, operation)
            default:
                throw CustomError.patchFailed
            }

            guard SHA256.hash(data: Data(content)).map({ String(format: "%02x", $0) }).joined() == file.sha256 else {
                throw CustomError.patchChecksumMismatch
            }

            let destination = directory.appendingPathComponent(file.href)
            try FileManager.default.createDirectory(
                at: destination.deletingLastPathComponent(),
                withIntermediateDirectories: true
            )
            try Data(content).write(to: destination)
        }
    }

    private static func decompress(_ input: [UInt8], prefix: [UInt8]?, expectedSize: Int?) throws -> [UInt8] {
        guard let context = ZSTD_createDCtx() else {
            throw CustomError.patchFailed
        }
        defer { ZSTD_freeDCtx(context) }

        guard ZSTD_isError(ZSTD_DCtx_setParameter(context, ZSTD_d_windowLogMax, windowLogMax)) == 0 else {
            throw CustomError.patchFailed
        }

        let size: Int
        if let expectedSize {
            size = expectedSize
        } else {
            let declared = input.withUnsafeBufferPointer { ZSTD_getFrameContentSize($0.baseAddress, $0.count) }
            guard declared != UInt64.max, declared != UInt64.max - 1, declared <= UInt64(Int.max) else {
                throw CustomError.patchFailed
            }
            size = Int(declared)
        }
        guard size >= 0 else {
            throw CustomError.patchFailed
        }

        var output = [UInt8](repeating: 0, count: size)
        let written: Int = try output.withUnsafeMutableBufferPointer { out in
            try input.withUnsafeBufferPointer { source in
                if let prefix {
                    let referenced = prefix.withUnsafeBufferPointer {
                        ZSTD_DCtx_refPrefix(context, $0.baseAddress, $0.count)
                    }
                    guard ZSTD_isError(referenced) == 0 else {
                        throw CustomError.patchFailed
                    }
                }
                return ZSTD_decompressDCtx(context, out.baseAddress, out.count, source.baseAddress, source.count)
            }
        }

        guard ZSTD_isError(written) == 0, written == size else {
            throw CustomError.patchFailed
        }
        return output
    }

    private static func readBase(_ base: URL, _ operation: [String: Any]) throws -> [UInt8] {
        guard let data = try? Data(contentsOf: base.appendingPathComponent(try safeHref(operation, "from"))) else {
            throw CustomError.patchFailed
        }
        return [UInt8](data)
    }

    private static func payload(_ payloads: [[UInt8]], _ operation: [String: Any]) throws -> [UInt8] {
        guard let index = operation["payload"] as? Int, index >= 0, index < payloads.count else {
            throw CustomError.patchFailed
        }
        return payloads[index]
    }

    private static func readLength(_ container: [UInt8], _ offset: inout Int) throws -> Int {
        guard offset + 4 <= container.count else {
            throw CustomError.patchFailed
        }
        var value = 0
        for byte in container[offset..<(offset + 4)] {
            value = (value << 8) | Int(byte)
        }
        offset += 4
        guard value >= 0, offset + value <= container.count else {
            throw CustomError.patchFailed
        }
        return value
    }

    private static func readBlock(_ container: [UInt8], _ offset: inout Int) throws -> [UInt8] {
        let length = try readLength(container, &offset)
        let block = Array(container[offset..<(offset + length)])
        offset += length
        return block
    }

    /// The plan is untrusted input, so its hrefs get the same relative-path rule as the manifest's.
    private static func safeHref(_ operation: [String: Any], _ key: String) throws -> String {
        guard let href = operation[key] as? String, !href.isEmpty, !href.hasPrefix("/") else {
            throw CustomError.patchFailed
        }
        for segment in href.split(separator: "/", omittingEmptySubsequences: false) {
            if segment.isEmpty || segment == "." || segment == ".." {
                throw CustomError.patchFailed
            }
        }
        return href
    }
}
